java-checks/src/main/java/org/sonar/java/checks/UnreachableCatchCheck.java

```diff
@@ -48,203 +42,67 @@
-          .collect(Collectors.toList());
-
-        if (hiddenTypes.size() == catchClause.types.size()) {
-          reportWholeCatchClause(catchClause);
-        } else {
-          for (Type hiddenType : hiddenTypes) {
-            reportSingleType(hiddenType);
-          }
-        }
-      }
+  private void checkWarnings(TryStatementTree tryStatementTree) {
+    Map<JWarning, TypeTree> exceptionTreeByWarning = new HashMap<>();
+    Map<TypeTree, Type> exceptionTypes = new HashMap<>();
+    for (CatchTree catchTree : tryStatementTree.catches()) {
+      caughtExceptionTypes(catchTree).stream()
+        .forEach(typeTree -> {
+          exceptionTypes.put(typeTree, typeTree.symbolType());
+          SyntaxToken typeFirstToken = typeTree.firstToken();
+          warnings.stream()
+            .filter(warning -> warning.contains(typeFirstToken))
+            .forEach(warning -> exceptionTreeByWarning.put(warning, typeTree));
+        });
     }
+    reportWarning(exceptionTreeByWarning, exceptionTypes);
+  }
 
-    void reportWholeCatchClause(CatchClauseInfo catchClause) {
-      List<Type> subtypesHiding = catchClause.types.stream()
-        .flatMap(type -> baseToSubtype.get(type).stream())
-        .filter(subtype -> isHiding(subtype, thrownTypes))
-        .collect(Collectors.toList());
-
-      reportIssue(catchClause.tree.catchKeyword(),
-        "Remove or refactor this catch clause because it is unreachable as hidden by previous catch blocks.",
-        subtypesHiding.stream().map(type -> new JavaFileScannerContext.Location("Already catch the exception", typeToTypeTree.get(type)))
-          .collect(Collectors.toList()),
-        null);
+  private static List<TypeTree> caughtExceptionTypes(CatchTree catchTree) {
+    TypeTree caughtException = catchTree.parameter().type();
+    if (caughtException.is(Tree.Kind.UNION_TYPE)) {
+      return ((UnionTypeTree) caughtException).typeAlternatives();
     }
+    return Collections.singletonList(caughtException);
+  }
 
-    void reportSingleType(Type hiddenType) {
-      List<Type> subtypesHiding = baseToSubtype.get(hiddenType).stream()
-        .filter(subtype -> isHiding(subtype, thrownTypes))
-        .collect(Collectors.toList());
-
-      reportIssue(typeToTypeTree.get(hiddenType),
-        "Remove this type because it is unreachable as hidden by previous catch blocks.",
-        subtypesHiding.stream().map(type -> new JavaFileScannerContext.Location("Already catch the exception", typeToTypeTree.get(type)))
-          .collect(Collectors.toList()),
+  private void reportWarning(Map<JWarning, TypeTree> exceptionTreeByWarning, Map<TypeTree, Type> exceptionTypes) {
+    for (Map.Entry<JWarning, TypeTree> treeByWarning : exceptionTreeByWarning.entrySet()) {
+      TypeTree reportTree = treeByWarning.getValue();
+      reportIssue(reportTree,
+        "Remove this type because it is unreachable, hidden by previous catch block(s).",
```

Comment:
I believe we should still provide different messages for a single type being unreachable vs. the whole catch block being unreachable. IIRC we introduced that distinction specifically because only talking about types confused some users.